barretenberg/cpp/src/barretenberg/vm2/tracegen/keccakf1600_trace.cpp

```diff
@@ -0,0 +1,782 @@
+                { C::keccak_memory_rw, rw ? 1 : 0 },
+                { C::keccak_memory_addr, addr + i },
+                { C::keccak_memory_space_id, event.space_id },
+                { C::keccak_memory_tag, static_cast<uint8_t>(tags[i]) },
+                { C::keccak_memory_tag_min_u64_inv,
+                  single_tag_errors[i]
+                      ? (FF(static_cast<uint8_t>(tags[i])) - FF(static_cast<uint8_t>(MemoryTag::U64))).invert()
+                      : 1 },
+                { C::keccak_memory_single_tag_error, single_tag_errors[i] ? 1 : 0 },
+                { C::keccak_memory_tag_error, tag_errors[i] ? 1 : 0 },
+            } });
+
+        // We get a "triangle" when shifting values to their columns from val00 bottom-up.
+        for (size_t j = i; j < num_rows; j++) {
+            trace.set(MEM_VAL_COLS[j - i], row, state_ff[j / 5][j % 5]);
+        }
+    }
+}
+
+} // namespace
+
+void KeccakF1600TraceBuilder::process_permutation(
+    const simulation::EventEmitterInterface<simulation::KeccakF1600Event>::Container& events, TraceContainer& trace)
+{
+    trace.set(C::keccakf1600_last, 0, 1);
+
+    uint32_t row = 1;
+    for (const auto& event : events) {
+        const bool out_of_range = event.src_out_of_range || event.dst_out_of_range;
+        const bool error = out_of_range || event.tag_error;
+
+        for (size_t round_idx = 0; round_idx < AVM_KECCAKF1600_NUM_ROUNDS; round_idx++) {
+            const auto& round_data = event.rounds[round_idx];
+
+            // Setting the selector, xor operation id, and operation id, round, round cst
+            trace.set(C::keccakf1600_sel, row, 1);
+            trace.set(C::keccakf1600_clk, row, event.execution_clk);
+            trace.set(C::keccakf1600_bitwise_xor_op_id, row, static_cast<uint8_t>(BitwiseOperation::XOR));
+            trace.set(C::keccakf1600_bitwise_and_op_id, row, static_cast<uint8_t>(BitwiseOperation::AND));
+            trace.set(C::keccakf1600_round, row, round_idx + 1); // round is 1-indexed
+            trace.set(C::keccakf1600_round_cst, row, simulation::keccak_round_constants[round_idx]);
+            trace.set(C::keccakf1600_thirty_two, row, AVM_MEMORY_NUM_BITS);
+
+            // Setting the rotation length constants
+            // If the constant is <= 32, we set it in the column.
+            // Otherwise, we set 64 - constant.
+            for (size_t k = 1; k < 25; k++) {
+                const size_t i = k / 5;
+                const size_t j = k % 5;
+                const auto rotation_len = simulation::keccak_rotation_len[i][j];
+                const auto value = rotation_len <= 32 ? rotation_len : 64 - rotation_len;
+                trace.set(RHO_ROTATION_LEN_COLS[k - 1], row, value);
+            }
+
+            // Selectors start and last.
+            // src_address required on first row
+            if (round_idx == 0) {
+                trace.set(C::keccakf1600_start, row, 1);
```

Comment:
In this file you have a mix of `trace.set(row, {{ }})` and single sets. I think we have mostly been using the bulk setter when we have to set many at the same time.